Lukeu/fix/more tests off runtime classpath

lukeu opened the pull request:
This change resolves the 'test resources in the runtime classpath' issue I was continuing to see following issue #354. I tested with my minimal example with Java 8 & 9, and my commercial workspace with Java 8 & both work as expected with this. However I really don't feel 'confident' with Eclipse or Buildship code yet, so I think this does need a good check - there seem to be quite a number of possibilities to consider.

(I spent some time trying to write an integration test for this based on RuntimeClasspathTest.groovy, but failed. I got as far as seeing that I'd need to construct a Jdt launch (rather than a Gradle launch) but that seemed to entail too much to set up, without some form of mocking or stubbing. There didn't seem to be something like that to emulate either?)

donat commented:
Hi! Thanks for investing time into creating a pull request.

I had to switch to a different project lately but I'm trying to get back to may Buildship duties. Expect replies to your previous questions in the upcoming days. 

Regarding the tests, it's indeed a tricky thing to implement. We do have [coverage to execute launch configurations and inspect the runtime classpath](https://github.com/eclipse/buildship/blob/6b651347ba2e529d516a4481825f7222c9d52768/org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/launch/ClasspathSeparationTest.groovy). Unfortunately it uses some Eclipse UI components, hence it is located in the `org.eclipse.ui.plugin`.

lukeu commented:
> I had to switch to a different project lately (...) expect replies to your previous questions in the upcoming days.

Thanks & sure, whenever makes sense - we humans are bad at multi-tasking.  :-)

lukeu commented:
Hi, I'm back from travels and came back to look at this.

> We do have coverage to execute launch configurations and inspect the runtime classpath.

Thanks for the pointer. Unfortunately I've hit some blockers and can't get the existing `ClasspathSeparationTest` to run: one test timed-out, and the rest failed because of a snapshot that is no longer [available here](https://services.gradle.org/distributions-snapshots/). After updating all the links to an available gradle-snapshot version, then all tests throw a timeout exception. Increasing the timeout 5s -> 60s did not help, something seems to be genuinely not working.

Would a way to move this forward be to just do a code review? The fix is essentially just a 2 line change that might just need a knowledgeable person to confirm that it makes sense. Otherwise someone with a working build could extend `createSampleProject` to split the test project into 2 projects, one that references the other. Then I imagine the existing checks would also cover this fix.

donat commented:
FYI I've fixed the `ClasspathSeparationTest`.

lukeu commented:
Just as a quick update: I've not had much time for "hobby coding" the past few weeks (time pressure at work & home - end of school term etc). I did pull this change down one evening but still had trouble - I still couldn't get the UI tests to execute properly locally :-(

I'm super keen to see this change merged however, as it requires an elaborate workaround which would be nice not to have to cherry-pick into each of my branches.

lukeu commented:
Thanks. I hope to fit in some time over the weekend to wrap these 2 PRs up.

lukeu commented:
Hi @donat, I've resolved my problem with the UI tests.

I've updated the 7 classpath separation tests to add coverage for the transient nature of the fix, as well as ensuring it works for resources (not just sources). I confirmed 1 of these tests failed before my fix and works afterwards, with JDK-8 and 9.

Testing notes:
 * Launching from Eclipse Oxygen, I get the same number of test failures before and after my fix (5 in core tests, 2 in the ui.view tests). I already fixed one by setting UTF-8 character encoding and another seems to be a locked file resource; so these both sound like issues with running on Windows vs Linux/Mac. Not sure about the rest.
 * I can't build/test using the gradle wrapper on the command-line, as these both give the following:

        [ant:taskdef] Could not load definitions from resource org/apache/maven/artifact/ant/antlib.xml. It could not be found.

Could this be a broken web-link, or perhaps because I don't have Maven installed?

donat commented:
Thanks .for the update, I'll take a look. 

If something unrelated fails please ignore it, at least in the context of this PR. Let's discuss them separately, in new issues. If you have time for the fix there, that's even better :)

Regarding the `antlib.xml`, the Buildship build itself relies on a local Maven repository that is created by the [scripts](https://github.com/eclipse/buildship/blob/master/buildSrc/src/main/groovy/eclipsebuild/mavenize/DeployMavenAntTaskExecutor.groovy#L32) in `buildSrc` which depends on `org.apache.maven:maven-ant-tasks`. Please let me know if you find out why it's not working for you.

donat commented:
@lukeu Thank you for the contribution!

lukeu commented:
Thanks a lot, very happy to see this one go in.
